common/lib/dependabot/dependency_group.rb

```diff
@@ -32,5 +60,81 @@ def to_config_yaml
         "groups" => { name => rules }
       }.to_yaml.delete_prefix("---\n")
     end
+
+    private
+
+    # TODO: Decouple pattern and exclude-pattern
+    #
+    # I think we'll probably want to permit someone to group by dependency type but still use exclusions?
+    #
+    # We probably need to think a lot more about validation to ensure we have _at least one_ positive-match rule
+    # out of pattern, dependency-type, etc, as well as `exclude-pattern` or we'll need to support it as an implicit
+    # "everything except exclude-patterns" if it can be configured on its own.
+    #
+    def matches_pattern?(dependency_name)
+      return true unless pattern_rules? # If no patterns are defined, we pass this check by default
+
+      positive_match = rules["patterns"].any? { |rule| WildcardMatcher.match?(rule, dependency_name) }
+      negative_match = rules["exclude-patterns"]&.any? { |rule| WildcardMatcher.match?(rule, dependency_name) }
+
+      positive_match && !negative_match
+    end
+
+    def matches_dependency_type?(dependency)
+      return true unless experimental_rules_enabled?
+      return true unless rules.key?("dependency-type") # If no dependency-type is set, match by default
+
+      rules["dependency-type"] == if dependency.production?
+                                    "production"
+                                  else
+                                    "development"
+                                  end
+    end
+
+    def pattern_rules?
+      rules.key?("patterns") && rules["patterns"]&.any?
+    end
+
+    def highest_semver_allowed
+      rules.fetch("highest-semver-allowed", DEFAULT_SEMVER_LEVEL)
+    end
+
+    def generate_ignore_condition!
+      return NullIgnoreCondition.new unless experimental_rules_enabled?
+
+      ignored_update_types = ignored_update_types_for_rules
+
+      return NullIgnoreCondition.new unless ignored_update_types.any?
+
+      Dependabot.logger.debug("The #{name} group has set ignores for update-type(s): #{ignored_update_types}")
+
+      Dependabot::Config::IgnoreCondition.new(
+        dependency_name: ANY_DEPENDENCY_NAME,
+        update_types: ignored_update_types
+      )
+    end
+
+    def ignored_update_types_for_rules
```

Comment:
Do we also foresee a need for folks to configure only updates for minor or major version bumps? Right now you'd always get everything _up to_ the specified range. It's a good starting point, but have we thought about the other scenario's?

Reply:
I originally started out with this, but this can be achieved using ignores, if the group had a highest semver value of `major`, you could add:

```
ignore:
  - dependency-name: *
     update-types: ["version-updates:semver-patch"]
```

The semantics of allowing something like "I only want patch-and-major updates in this group" is hard to reason about if we make it too broad, so I'd ended up leaning on our existing configuration options to keep the group semantics simple.

Reply:
I've been thinking about this more and implementing as a list might end up being the right answer if someone wants the following:

```
groups
  patches:
    update-types: ["patches"]
  small-bumps:
    update-types: ["minor"]
```

With the `highest-semver-allowed` approach, this would _almost_ work, but it would be surprising if you only had patch versions available for one or more dependencies - you end up with the `patches` and `small-bumps` PRs containing both containing the patch version.